src/NHibernate/Mapping/Constraint.cs

```diff
@@ -38,6 +39,142 @@ public IEnumerable<Column> ColumnIterator
 			get { return columns; }
 		}
 
+		/// <summary>
+		/// Generate a name hopefully unique using the table and column names.
+		/// Static so the name can be generated prior to creating the Constraint.
+		/// They're cached, keyed by name, in multiple locations.
+		/// </summary>
+		/// <param name="prefix">A name prefix for the generated name.</param>
+		/// <param name="table">The table for which the name is generated.</param>
+		/// <param name="referencedTable">The referenced table, if any.</param>
+		/// <param name="columns">The columns for which the name is generated.</param>
+		/// <returns>The generated name.</returns>
+		/// <remarks>Hybrid of Hibernate <c>Constraint.generateName</c> and
+		/// <c>NamingHelper.generateHashedFkName</c>.</remarks>
+		public static string GenerateName(
+			string prefix, Table table, Table referencedTable, IEnumerable<Column> columns)
+		{
+			// Use a concatenation that guarantees uniqueness, even if identical names
+			// exist between all table and column identifiers.
+			var sb = new StringBuilder("table`" + table.Name + "`");
+			if (referencedTable != null)
+				sb.Append("references`" + referencedTable.Name + "`");
+
+			// Ensure a consistent ordering of columns, regardless of the order
+			// they were bound.
+			// Clone the list, as sometimes a set of order-dependent Column
+			// bindings are given.
+			var alphabeticalColumns = new List<Column>(columns);
+			alphabeticalColumns.Sort(ColumnComparator.Instance);
+			foreach (var column in alphabeticalColumns)
+			{
+				var columnName = column == null ? "" : column.Name;
+				sb.Append("column`").Append(columnName).Append("`");
+			}
+			// Hash the generated name for avoiding collisions with user choosen names.
+			// This is not 100% reliable, as hashing may still have a chance of generating
+			// collisions.
+			var name = prefix + HashName(sb.ToString());
+
+			// Hibernate uses an algorithm yielding names shorter than 30 characters. But we cannot
+			// use it (see HashName). And also we have DB limited to even less (Informix)...
+			if (name.Length > 30)
+			{
+				// This, of course, increases the collision risk.
+				name = name.Substring(0, 30);
+			}
+
+			return name;
+		}
+
+		#region Name generation support methods
+
+		/// <summary>
+		/// Hash a constraint name. Convert the hash digest to base 32
+		/// (full alphanumeric) for shortening the hash string representation
+		/// while keeping it suitable for db names.
+		/// </summary>
+		/// <param name="name">The name to be hashed.</param>
+		/// <returns>The hased name.</returns>
+		private static string HashName(string name)
+		{
+			// Hibernate uses MD5, but with .Net this would throw on FIPS enabled machine.
```

Comment:
This thing is from my point of view a very unfortunate design decision of the .Net Framework: if FIPS is enabled on the machine, any usage of non FIPS compliant algorithm is forbidden. But FIPS does not forbid them, it just forbids to use them for handling a security related task. And here we are not handling a security related task. But there is no way to tell it for allowing using MD5 here even on FIPS enabled machine. See [this Microsoft blog](https://blogs.technet.microsoft.com/secguide/2014/04/07/why-were-not-recommending-fips-mode-anymore/) (may be very long to load) for more information.

So if we want to use MD5 here without causing NHIbernate to be unusable on a FIPS enabled machine, we would have to use a custom implementation, like [this one](https://rosettacode.org/wiki/MD5/Implementation#C.23).